TEG refactor

Markolie opened the pull request:
Refactors our TEG so it's up-to-date with /tg/'s latest version. No real functional changes, except that the TEG now actually works, and the cold/heat circulation side of the TEG can now be adjusted using a multitool. 

This is largely untested, as I'm not familiar enough with the TEG to determine whether or not it's working well enough. The code works, the TEG works, I just don't know if it works well. If anyone's willing to test if for me, let me know and I can host a server for you. This shouldn't be merged before then.

Maybe fixes https://github.com/ParadiseSS13/Paradise/issues/2860.
Maybe fixes https://github.com/ParadiseSS13/Paradise/issues/1850.

🆑 Markolie
bugfix: Thermo-electric generators (TEG's) are once again fully functional. 
tweak: The cold/heat side of a TEG can now be adjusted using a multitool.
/🆑

Markolie commented on code/game/machinery/pipe/construction.dm:
```diff
@@ -150,10 +151,16 @@
+		
+		var/obj/machinery/atmospherics/binary/circulator/circP = make_from
+		if(istype(circP) && circP.side == circP.CIRC_RIGHT)
+			src.flipped = 1
```
I had to "snowflake" this in here, but considering construction.dm is already a giant turd I don't realy care.

FalseIncarnate replied:
.src

Markolie replied:
You should have a look at the rest of the file :P. I'm keeping it here for consistency's sake.

Fox-McCloud commented on code/ATMOSPHERICS/components/binary_devices/circulator.dm:
```diff
@@ -3,76 +3,68 @@
+	can_unwrench = 1
+	
 /obj/machinery/atmospherics/binary/circulator/New()
+	initialize()
```
I believe `initialize` is called in `New` somewhere up the tree--- @Crazylemon64 would know better.

Markolie replied:
Ah, I put that there for testing purposes (so I could spawn the circulators with the game panel). I'll get rid of it.

Taaketa commented:
I would gladly test this. I love the TEG and would love to see it work on Paradise.

lordpidey commented:
>It's not really possible to have the pipes delivered without rewriting parts of the cargo code (just for these pipes) due to unwrenched pipes all being /obj/item/pipe's.

Couldn't you just make a subclass of /obj/item/pipe with the appropriate variables set, and put those subclasses in the cargo crate?

Markolie commented:
The variables are directly inserted into new as parameters, so unfortunately enough not.

Markolie commented:
Looks like it was very much possible to have cargo deliver the circulators - that's now the case again. It's now also possible to change the direction of the cold/heat circ side with a multitool.

Taaketa commented:
@Markolie 

I've tried a compile and clean compile but the .dmb file seems to not want to appear on your branch. Have I done something wrong?

Markolie commented:
@Taaketa I'm not sure why it'd do that. Perhaps try downloading a .zip of the repository, unpacking that and then try to compile it?

Taaketa commented:
@Markolie Apparently third times a charm, compiling, clean compile didn't work and then clicking Run - Compile did. Randomness.

lordpidey commented:
>The variables are directly inserted into new as parameters, so unfortunately enough not. I'll have another look though.

I might be able to do this for you this weekend.  Send me  a reminder on Friday.

Markolie commented:
@lordpidey Ah - I figured out a way to do it. Appreciate the offer though!

Fox-McCloud commented:
Let us know when this is tested.

Taaketa commented:
@Markolie

Unfortunately I cannot seem to get the TEG to play ball at all.

I've tried several different setups. One genuine setup and then 3 adminbused setups and the circulators do not seem to be picking up any gas. Tried, wrenching, unwrenching. Wrenching with gas being pushed, unwrenching when emptied, wrenching before laying pipe, unwrenching after laying pipe. Wrenching with no SMES connected, unwrenching during gas pumping and no smes connected. But the TEG doesn't want to push gas through.

Screenshots:

http://imgur.com/a/omupP


Also some additional notes I noticed whilst messing with the TEG.

- Mtool only works once all parts are in place and must be used on the TEG Computer and not the circulators. (instructional information rather than a bug)

- TEG doesn't state what is input or output ports for respective Circulators in it's description, just which side is which. (Both Circulators appear to accept input only from the south.)

- Circulators cannot be rotated. Flip works rotates 180 degrees. Alt Click does nothing.

- TEG console cannot be rotated and has to be pushed to be aligned correctly. Alt Click does nothing.

- 2 circulators show up on the game panel.

/obj/machinery/atmospherics/binary/circulator (no sprite)

/obj/item/pipe/circulator (sprite)

Markolie commented:
Thank you for taking the time to test the implementation @Taaketa - it's much appreciated. I'll have the fixes up by tonight hopefully.

To respond:
- I'm guessing the TEG stopped producing power due to one of my last two changes - whoops! I figured it'd properly rotate the air still, but I messed up somewhere.
- I'll add the input and output ports to the circulator description.
- Circulators not being rotated is intentional - the circulator/TEG can only be set up horizontally. I'll see if I can make the rotate command flip the circulators instead.
- The circulator pipe item is used when the circulator is deconstructed, and once it's constructed it turns into the machine. Generally speaking you should always spawn the pipe item and wrench that.

Thanks again for the testing - it's much appreciated.

Markolie commented:
And it turns out I forgot to re-add the generator to the processing_machines, so it wasn't processing at all. I've fixed that issue, added the descriptions about the input/output port to the connectors and made the connectors rotate when they get flipped.

I've done some more manual testing, and I'm 80% sure it's working well. Would appreciate another test though!

Taaketa commented:
@Markolie  Same link as before to download?

Markolie commented:
That's correct. Thank you!

Taaketa commented:
It appears to be functioning correctly. Some observations.

- The TEG generator still spawns vertically and has to be pushed to make it rotate to the horizontal alignment. No flip or rotate available when right clicking.

- The TEG can seemingly only generate a maximum of 500,000 watts. Regardless of Temperature differences of incoming gas. Not sure if a deliberate limitation set into it or some kind of bug. (Screen shots below)


http://imgur.com/a/Z68b7

- SMES cells do not seem to accept TEG power but this maybe a bug with the SMES cells. The wire under the TEG was powered when checking with an Mtool.

Markolie commented:
Thanks once more!
- I'm not sure what you mean by the TEG generator spawning vertically. Do you have a screenshot of that?
- There's no hard cap on the TEG, so it really depends on the mixes/temperature. It may well be the most you can get out of it though - not sure.
- The SMES cells not taking the TEG power is most likely not a TEG bug (as it uses generic procs to add power to the network). Perhaps it's because the SMES is on Z2?

Either way, I think this is ready for merging. If the vertical TEG generator issue does turn out to be a bug I'll make sure to fix that.

Taaketa commented:
Here are screenshots.

First shows it spawned in.

Second shows it pushed to the right.

Third shows that even in vertical "alignment" it still reads the circulators ok. I got confused with the old teg which could be rotated and it matter what direction the grey bars were in.

http://imgur.com/a/i8SmB

Markolie commented:
Ah, like that. That must be a relic of the past - the direction of the TEG no longer matters. It'll connect to the circulators regardless of it's dir.

Markolie commented:
Ready to go @Fox-McCloud.

Spacemanspark commented:
Merge when
